File: LeetCodeProblemsGo/LeetCode797AllPathsFromSourceToTarget/LeetCode797AllPathsFromSourceToTarget.go
package LeetCode797AllPathsFromSourceToTarget

func dfs(graph [][]int, length int, visited []bool, result *[][]int, path *[]int, node int) {
	if node == length-1 {
		newPath := make([]int, len(*path))
		copy(newPath, *path)
		*result = append(*result, newPath)
		return
	}
	for _, neighbour := range graph[node] {
		if visited[neighbour] {
			continue
		}
		visited[neighbour] = true
		*path = append(*path, neighbour)
		dfs(graph, length, visited, result, path, neighbour)
		*path = (*path)[:len(*path)-1]
		visited[neighbour] = false
	}
}

func allPathsSourceTarget(graph [][]int) [][]int {
	length := len(graph)
	visited := make([]bool, length)
	visited[0] = true
	path := []int{0}
	result := [][]int{}
	dfs(graph, length, visited, &result, &path, 0)
	return result
}

File: LeetCodeProblemsJava/src/Problems/LeetCode797AllPathsFromSourcetoTarget.java
package Problems;

import java.util.ArrayList;
import java.util.List;

public class LeetCode797AllPathsFromSourcetoTarget {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
       int length = graph.length;
       List<List<Integer>> result = new ArrayList<>();
       boolean[] visited = new boolean[length];
       List<Integer> path = new ArrayList<>(List.of(0));
       dfs(graph,result,path,length,visited,0);
       return result;
    }
    void dfs(int[][] graph,List<List<Integer>> res,List<Integer> currentPath,int length,boolean[] visited,int node){
        if(node==length-1){
            List<Integer> newPath = new ArrayList<>(currentPath);
            res.add(newPath);
            return;
        }
        for (int neighbour:graph[node]){
            if(visited[neighbour])
                continue;
            visited[neighbour]=true;
            currentPath.add(neighbour);
            dfs(graph,res,currentPath,length,visited,neighbour);
            currentPath.remove(Integer.valueOf(neighbour));
            visited[neighbour]=false;
        }


    }
}
